Tree view interface guidelines (#285)
primer/design: content/components/treeview.mdx

## Nesting

Nodes are visually nested using whitespace and alignment.

### Align leading icons

<Box as="figure" display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} m={0} sx={{gap: 3}}>
  <Box flex={1}>
    Icons at the same level are center-aligned. The chevron should not break the alignment: parent nodes and child nodes
    keep their leading visuals aligned.
  </Box>
  <img
    width="456"
    role="presentation"
    alt="19 leadingVizAlignment"
    src="https://user-images.githubusercontent.com/2313998/187768704-08fac391-f331-4a47-9181-073f1360677c.png"
  />
</Box>

### Indentation

<Box as="figure" display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} m={0} sx={{gap: 3}}>
  <Box flex={1}>Each level of nesting increments the size of the indentation.</Box>
  <img
    width="456"
    role="presentation"
    alt="20 nodesWithIndentations"
    src="https://user-images.githubusercontent.com/2313998/187768706-1910025a-f69b-487f-987b-319f3022cd09.png"
  />
</Box>

### Nesting level indicator lines

> In devices with `:hover` support, the nesting indicator lines fade in when the user mouses over the entire component, or when there's focus inside the component.

Reviewer: ```suggestion
On devices with `:hover` support, the nesting indicator lines fade in when the user mouses over the entire component, or when there's focus inside the component.
```

    src="https://user-images.githubusercontent.com/2313998/187768708-805527f9-3b7c-4f6b-ae37-e232f83916b6.png"
  />
  <img
    width="456"
    role="presentation"
    alt="21 2 alwaysShowLinesOnTouch"
    src="https://user-images.githubusercontent.com/2313998/187768711-2702a7aa-e902-4568-ab99-4570e28bf646.png"
  />
</Box>

### Handling deeply nested nodes

Navigating through deeply nested nodes can be cumbersome and visually clunky. If accessing nodes deeper than 10 levels deep is a common interaction for your use-case, reconsider whether a tree view is the best pattern.

Tree views are designed to be horizontally compact in order to support many levels of nesting and long node labels. This makes content truncation and horizontal overflow less likely to occur for deeply nested nodes, but infinite nesting makes those cases impossible to prevent entirely.

There are some strategies to avoid a tree view that horizontally overflows it's container:

- Allow the tree view's container to be manually resized
- Once a certain depth is reached, start the tree view from a deeper parent node and provide a way to navigate back up the tree
- Remove leading and trailing visuals

> If a tree view is truly the best pattern for your use case and none of the suggested strategies prevent horizontal overflow, opt for horizontal scrolling as a last resort instead of hiding overflowing content.

Reviewer: ```suggestion
If a tree view is truly the best pattern for your use-case and none of the suggested strategies prevent horizontal overflow, opt for horizontal scrolling as a last resort instead of hiding overflowing content.
```

Reviewer: You used the hyphenated version above already. So we should probably stick to one.

#### Typeahead behavior

<Box display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} sx={{gap: 3}}>
  <Box flex={1}>
    <p>
      Focus moves to the next node with a name that starts with the typed character(s). Wait for 300ms from the last key
      press to stop searching.
    </p>
    <p>
      Typeahead behavior is case insensitive, and any printable character can be used, not just alphanumeric characters.
      The exception is the “*” character, which is used to expand all nodes.
    </p>
    <p>
      The `Space` key is ignored because it's being reserved for toggling a tree view node's checkbox. Nodes with
      checkboxes are not supported yet, but they may in the future.
    </p>
  </Box>
  <CustomVideoPlayer
    width="456"
    loop
    src="https://user-images.githubusercontent.com/2313998/187719392-3693e45a-f108-4b78-84ef-0cfe4c7067f9.mp4"
  />
</Box>

### Focus behavior

<Box
  display="flex"
  alignItems="center"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  sx={{gap: 3}}
>

<Box flex={1}>

#### Focus in

> If I'm moving focus into the TreeView for the first time, focus on the first node.
> 
> If I'm moving focus back into the TreeView after I've already interacted with it, focus the previously focused TreeView node.

Reviewer: ```suggestion
If moving focus into the TreeView for the first time, focus on the first node.

If moving focus back into the TreeView after the user already interacted with it, focus the previously focused TreeView node.
```

Reviewer: You never use this form, so it feels a bit foreign here.

<Box sx={{fontSize: 3}} class="lead" as="p">
  A tree view is a hierarchical list of items that may have a parent-child relationship where children can be toggled
  into view by expanding or collapsing their parent item.
</Box>

## Anatomy

<img
  width="960"
  role="presentation"
  src="https://user-images.githubusercontent.com/2313998/187719332-3ef9aa9e-476f-4e5a-93fa-5fbd203ece4a.png"
/>

1. **Parent node**: A node that hides and shows child nodes by expanding and collapsing
2. **End node**: A node with no children
3. **Nesting level indicator lines**: Each line represents a level of nesting depth
4. **Chevron**: Toggles a parent node expanded or collapsed and indicates expanded/collapsed state
5. **Leading visual (optional)**: A visual cue for additional context
6. **Node label**: The text label for the node
7. **Trailing visual (optional)**: Same as a leading visual, but at the end

## Content

### Node label

Node labels should be a succinct title for whatever entity the node represents. For example, a file name.

By default, labels are truncated to a single line. However, node labels may be wrapped to multiple lines in cases where the entire label must always be visible.

> When a node's text is truncated, it should still be accessible to users via a browser-native tooltip.

Reviewer 2: @mperrotti I think one issue can be that most browsers will not display the browser tooltip on focus, if I remember right 😞 Here's a post from Scott O'Hara that I appreciated on this topic: https://www.24a11y.com/2017/the-trials-and-tribulations-of-the-title-attribute/ (although it might be outdated at this point lol)

> When a node's text is truncated, it should still be accessible to users via a browser-native tooltip.

Author: Thank you! The React tooltip also doesn't show when focusing the target, nor does it require the target be focusable 😭

https://primer.style/react/Tooltip

> When a node's text is truncated, it should still be accessible to users via a browser-native tooltip.

Reviewer 3: Could we wrap the Radix Tooltip component? https://www.radix-ui.com/docs/primitives/components/tooltip

> When a node's text is truncated, it should still be accessible to users via a browser-native tooltip.

Author: @colebemis - we could if other folks are comfortable with that and it's accessible

> When a node's text is truncated, it should still be accessible to users via a browser-native tooltip.

Reviewer 4: @colebemis, @mperrotti I'm cool with this!

    />
    <Caption role="presentation" color="fg.subtle">
      Wrapping node label text
    </Caption>
  </Box>
</Box>

### Leading and trailing visuals

Leading and trailing visuals may be used as visual cues that communicate more information about a node.

If the leading or trailing visual contains useful information, make sure it has a screen-reader accessible text label.

Either all nodes have leading visuals, or no nodes have leading visuals.

<DoDontContainer>
  <Do>
    <img
      width="456"
      role="presentation"
      alt="3 1 DO-consistentLeadingViz"
      src="https://user-images.githubusercontent.com/2313998/187719338-9f860dfa-2110-4cf7-8d3b-583ff569605f.png"
    />
    <Caption role="presentation" color="fg.subtle">
      Render all nodes with leading visuals or as text-only
    </Caption>
  </Do>
  <Dont>
    <img
      width="456"
      role="presentation"
      alt="3 2 DONT-mixLeadingViz"
      src="https://user-images.githubusercontent.com/2313998/187719339-83fd7065-5b85-427f-8650-7fc0cae7214d.png"
    />
    <Caption role="presentation" color="fg.subtle">

>       Don't have a tree where only some nodes have leading visuals

Author: The node label is not marked as optional, but I could explicitly say that here if you don't think that's clear enough.

>       Don't have a tree where only some nodes have leading visuals

Author: I added more explicit guidance about text labels being required.

>       Don't have a tree where only some nodes have leading visuals

Reviewer 4: Nice!

    </Caption>
  </Dont>
</DoDontContainer>

## Interaction

> Keyboard users can expand or collapse nodes without activating them by using the left and right arrow keys.

Author: I think you meant <kbd>Enter</kbd> and <kbd>Space</kbd>, but I got guidance from @jscholes that <kbd>Space</kbd> should be reserved for checking/unchecking nodes.

> Keyboard users can expand or collapse nodes without activating them by using the left and right arrow keys.

Reviewer 4: Yes and yes. Confirmed via a working session at 1:00pm today.

<Box display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} sx={{gap: 3}}>
  <img
    width="456"
    role="presentation"
    alt="21 1 showLinesOnHover"
    src="https://user-images.githubusercontent.com/2313998/187768708-805527f9-3b7c-4f6b-ae37-e232f83916b6.png"
  />
  <img
    width="456"
    role="presentation"
    alt="21 2 alwaysShowLinesOnTouch"
    src="https://user-images.githubusercontent.com/2313998/187768711-2702a7aa-e902-4568-ab99-4570e28bf646.png"
  />
</Box>

### Handling deeply nested nodes

Navigating through deeply nested nodes can be cumbersome and visually clunky. If accessing nodes deeper than 10 levels deep is a common interaction for your use-case, reconsider whether a tree view is the best pattern.

Tree views are designed to be horizontally compact in order to support many levels of nesting and long node labels. This makes content truncation and horizontal overflow less likely to occur for deeply nested nodes, but infinite nesting makes those cases impossible to prevent entirely.

There are some strategies to avoid a tree view that horizontally overflows it's container:

> - Allow the tree view's container to be manually resized

Author: This is not functionality that will be a part of the tree view component, so I don't think this needs to be documented in the TreeView API specs.

None of our components support this behavior _yet_, but there has been talk about supporting it in [SplitPageLayout](https://primer.style/react/SplitPageLayout).

    <li>an FAQ that collapses answers under question headings</li>
  </Box>
</Box>

</Box>

<DoDontContainer>
  <Do>
    <img
      width="456"
      role="presentation"
      alt="7 1 DO-useForList"
      src="https://user-images.githubusercontent.com/2313998/187719349-c028a00f-0a27-4be2-bf67-815c2a036916.png"
    />
    <Caption color="fg.subtle">Use a tree view for a list</Caption>
  </Do>
  <Dont>
    <img
      width="456"
      role="presentation"
      alt="7 2 DONT-useForAccordion"
      src="https://user-images.githubusercontent.com/2313998/187719350-de23e9ac-fd10-4e56-b2a4-aca93255adb7.png"
    />
    <Caption color="fg.subtle">Don't use a tree view for expanding and collapsing sections of content</Caption>
  </Dont>
</DoDontContainer>

### Expanding and collapsing nodes

A tree view can be a frustrating pattern if it forces users to spend a lot of time expanding and collapsing nodes to find what they're looking for.

#### Expanded by default

<Box display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} sx={{gap: 3}}>
  <figcaption>
    <p>When a child node is active, all of it's parent nodes should be expanded.</p>

>     <p>If it's likely that a user will want to interact with all or most of the nodes in the tree, render the tree with all parent nodes expanded to start.</p>

Author: One example would be the "Files changed" tab of a PR. We start with all parent nodes expanded to make it easier for reviewers to scan and navigate the changed files.

>     <p>If it's likely that a user will want to interact with all or most of the nodes in the tree, render the tree with all parent nodes expanded to start.</p>

Author: Added an example to the guidelines. If you think we need even more clarity, let me know.

>     <p>If it's likely that a user will want to interact with all or most of the nodes in the tree, render the tree with all parent nodes expanded to start.</p>

Reviewer 4: Ah, I gotcha. Your example makes sense to me, so I'll update guidance on my end of things.

  <Do>
    <img
      width="456"
      role="presentation"
      alt="18 1 DO-renderWithEnoughSpace"
      src="https://user-images.githubusercontent.com/2313998/187768699-d97f4119-f989-487c-9d81-aa9253dbd580.png"
    />
    <Caption color="fg.subtle">
      Render a tree view in an area with enough vertical space to comfortably browse the list, and doesn't push
      important content or controls below the bottom of the viewport
    </Caption>
  </Do>
  <Dont>
    <img
      width="456"
      role="presentation"
      alt="18 2 DONT-renderTooSmall"
      src="https://user-images.githubusercontent.com/2313998/187768702-2881810c-7d1d-4e36-ae85-dfc35cf95cda.png"
    />
    <Caption color="fg.subtle">
      Don't render a tree view in an area that pushes down other content or forces users to scroll through small
      segments of the list
    </Caption>
  </Dont>
</DoDontContainer>

## Accessibility

### Label

<Box m={0}>
  <Box as="p">
    Tree views need an accessible name to be supplied, ideally via `aria-labelledby` pointing at an appropriate heading
    before the tree. If the tree view has enough surrounding context that it doesn't need a visible label, `aria-label`

>     may be used instead.

Author: I didn't realize there were problems with `aria-label`! Could you explain the assistive tech support issues or point me to some resources that could explain them?

| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>a-z</Box>, <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>A-Z</Box>, all printable characters | <ul><li>Focus moves to the next node with a name that starts with the typed character.</li><li>Search wraps to first node if a matching name is not found among the nodes that follow the focused node.</li><li>Search ignores nodes that are descendants of closed nodes.</li><li>Focus should not move if no nodes match the search.</li></ul> |

#### Typeahead behavior

<Box display="flex" flexDirection={['column', 'column', 'column', 'column', 'row']} sx={{gap: 3}}>
  <Box flex={1}>
    <p>

>       Focus moves to the next node with a name that starts with the typed character(s). Wait for 300ms from the last key

Author: I don't feel strongly either. I say let's generously give users a whole extra `50ms` to type and go with `300ms` 😂

>       Focus moves to the next node with a name that starts with the typed character(s). Wait for 300ms from the last key

Reviewer 4: Deal!

      press to stop searching.
    </p>
    <p>
      Typeahead behavior is case insensitive, and any printable character can be used, not just alphanumeric characters.

>       The exception is the “*” character, which is used to expand all nodes.

Author: Oops - missed this line. Thanks for catching this!

    </p>
    <p>
      The `Space` key is ignored because it's being reserved for toggling a tree view node's checkbox. Nodes with

>       checkboxes are not supported yet, but they may in the future.

Reviewer 4: This conflicts with [proposed documentation in the TreeView API documentation](https://github.com/github/primer/pull/1273/files#diff-6e09f5204050a1c218cd30c0b90bd87a09d2a872eb4b4a56e771e56ae50eb793R19). 

@jscholes: I'd be curious for your opinion here. Should we proactively remove <kbd>Space</kbd> functionality in anticipation of this upcoming functionality?

Reviewer 4: This is no longer a question, resolved in a conversation today.
